Add Git and GitHub section to SOP (#52)
emlab-ucsb/SOP: 04-gh/03-commits.qmd

## Atomic commits

A commit should represent one logical change. This principle is sometimes called making commits *atomic*. An atomic commit might be: adding a new function, fixing a specific bug, renaming a variable for clarity, or updating the README. A single commit should *not* include multiple unrelated changes like cleaning data, fitting a model, generating three figures, and updating the README all in one.

Atomic commits have practical benefits. When something breaks, a focused commit history makes it much easier to identify which change introduced the problem. When a collaborator reviews your work, smaller and more targeted commits are easier to understand. When you need to undo a change, you can revert a specific commit without also undoing unrelated work.

## Writing good commit messages

The commit message is addressed to a reader -- a future collaborator, a reviewer, or yourself in six months -- who is trying to understand why a change was made. The *what* is visible in the diff. The message should explain the *why*.

A commit message has two parts: a subject line and an optional body.

> **The subject line** should be a concise summary of the change, written in the imperative mood (as though giving an instruction), and no longer than 72 characters. For example:

Reviewer: ```suggestion
**The subject line** should be a concise summary of the change. It should be capitalized, written in the imperative mood (as though giving an instruction), ended without a period, and no longer than 50 characters.^[If you wish, see https://cbea.ms/git-commit/ for an explanation of the reasoning behind these rules.]  For example:
```

```

> Filter observations with missing species ID before modeling

Reviewer: ```suggestion
Filter observations with missing species ID
```

```

Rather than:

```
fixed the thing
updated script
more changes
```

**The body**, if needed, provides context: what motivated the change, what alternatives were considered, or what side effects to be aware of. Separate the body from the subject with a blank line.

A full example:

```

> Filter observations with missing species ID before modeling

Reviewer: ```suggestion
Filter observations with missing species ID
```
